core/domain/story_domain.py

```diff
@@ -0,0 +1,369 @@
+        self.last_updated = last_updated
+        self.version = version
+
+    def to_dict(self):
+        """Returns a dict representing this Story domain object.
+
+        Returns:
+            A dict, mapping all fields of Story instance.
+        """
+        return {
+            'id': self.id,
+            'title': self.title,
+            'description': self.description,
+            'notes': self.notes,
+            'language_code': self.language_code,
+            'schema_version': self.schema_version,
+            'version': self.version,
+            'story_nodes': [
+                node.to_dict() for node in self.story_nodes
+            ]
+        }
+
+    @classmethod
+    def from_dict(
+            cls, story_dict,
+            story_created_on=None, story_last_updated=None):
+        """Return a Story domain object from a dict.
+
+        Args:
+            story_dict: dict. The dictionary representation of the
+                story.
+            story_created_on: datetime.datetime. Date and time when the
+                story is created.
+            story_last_updated: datetime.datetime. Date and time when
+                the story was last updated.
+
+        Returns:
+            Story. The corresponding Story domain object.
+        """
+        story = cls(
+            story_dict['id'], story_dict['title'],
+            story_dict['description'], story_dict['notes'],
+            [
+                StoryNode.from_dict(node_dict)
+                for node_dict in story_dict['story_nodes']
+            ], story_dict['schema_version'], story_dict['language_code'],
+            story_dict['version'], story_created_on, story_last_updated)
+
+        return story
+
+    @classmethod
+    def create_default_story(
+            cls, story_id, title=feconf.DEFAULT_STORY_TITLE,
+            description=feconf.DEFAULT_STORY_DESCRIPTION,
+            notes=feconf.DEFAULT_STORY_NOTES,
+            language_code=constants.DEFAULT_LANGUAGE_CODE):
+        """Returns a story domain object with default values.
+
+        Args:
+            story_id: str. The unique id of the story.
+            title: str. The title of the story.
+            category: str. The category of the story.
+            objective: str. The objective of the story.
+            language_code: str. The language code of the story (like 'en'
+                for English).
+
+        Returns:
+            Story. The Story domain object with the default values.
+        """
+        return cls(
+            story_id, title, description, notes, [],
+            feconf.CURRENT_STORY_SCHEMA_VERSION, language_code, 0)
+
+    @classmethod
+    def update_story_contents_from_model(
+            cls, versioned_story_contents, current_version):
+        """Converts the states blob contained in the given
+        versioned_story_contents dict from current_version to
+        current_version + 1. Note that the versioned_story_contents being
```

Comment:
This seems a bit odd. Consider the scenario where the feconf.CURRENT_STORY_SCHEMA_VERSION is 5 and we've just retrieved a datastore object that's got a dictblob with schema version 2.

This method should migrate that dictblob one version at a time from 2 --> 3 --> 4 --> 5. It doesn't seem to be doing that?

You also have two sources of truth for current_version (the arg that's passed in and feconf.CURRENT_STORY_SCHEMA_VERSION) and I'm a bit confused why that's the case. The latter should suffice on its own?

Reply:
Yeah, I mentioned about that in my previous replies. current_version is actually the present version of that story (should I rename it?). Also, this whole function is called in a loop from story_services.

Reply:
feconf.CURRENT_STORY_SCHEMA_VERSION is just mentioned here as a final check against version going beyond it. A check is present in story_services, so it is not required to have feconf.CURRENT_STORY_SCHEMA_VERSION here. Shall I remove that then?

Reply:
Probably yes, though this kind of suggests moving both things into story_domain (and possibly consolidating into one method) so that there's no danger of wrong arguments being passed to the second method.

Reply:
Update from quick discussion with @BenHenning -- let's just follow the similar structure in exp_domain/exp_services for now. If we do any refactoring, let's do it for all objects at the same time and not just these ones.

Reply:
Ah, sorry, didn't realize that. I think we really need to refactor the migration stuff so that it works the same way for everything...

For now I think it is fine to follow exp_domain (per earlier comments). I'll work with @BenHenning on thinking through the more general refactor.